File: Presentation/Styling/MarkdownRenderer.swift
import Cocoa

/// Presentation: Markdown Renderer
/// Applies markdown styling to NSTextStorage
class MarkdownRenderer {

    enum Style {
        case notes      // Larger headers for notes view
        case analysis   // Smaller headers for analysis view

        var headerFontSize: CGFloat {
            switch self {
            case .notes: return 22
            case .analysis: return 18
            }
        }

        var subHeaderFontSize: CGFloat {
            switch self {
            case .notes: return 18
            case .analysis: return 16
            }
        }

        var bodyFontSize: CGFloat {
            switch self {
            case .notes: return 14
            case .analysis: return 13
            }
        }

        var codeFontSize: CGFloat {
            switch self {
            case .notes: return 13
            case .analysis: return 12
            }
        }
    }

    private let style: Style
    private let syntaxHighlighter: SyntaxHighlighter
    private let yellowColor = NSColor(red: 1.0, green: 0.84, blue: 0.0, alpha: 1.0)
    private let headerColor = NSColor(red: 1.0, green: 0.84, blue: 0.0, alpha: 1.0)
    private let codeColor = NSColor(red: 0.90, green: 0.92, blue: 0.95, alpha: 1.0)
    private let codeBackgroundColor = NSColor(red: 0.12, green: 0.13, blue: 0.15, alpha: 1.0)
    private let inlineCodeColor = NSColor(red: 0.98, green: 0.72, blue: 0.60, alpha: 1.0)
    private let inlineCodeBgColor = NSColor(red: 0.15, green: 0.15, blue: 0.20, alpha: 0.8)
    private let linkColor = NSColor(red: 0.4, green: 0.7, blue: 1.0, alpha: 1.0)
    private let markerColor = NSColor(white: 0.5, alpha: 0.6)

    init(style: Style = .notes) {
        self.style = style
        self.syntaxHighlighter = SyntaxHighlighter(fontSize: style.codeFontSize)
    }

    /// Render markdown in the given text storage
    func render(in storage: NSTextStorage) {
        let text = storage.string

        storage.beginEditing()

        // Reset formatting - clean white text with monospace font
        let baseAttributes: [NSAttributedString.Key: Any] = [
            .font: NSFont.monospacedSystemFont(ofSize: style.bodyFontSize, weight: .regular),
            .foregroundColor: NSColor(white: 0.95, alpha: 1.0)
        ]
        storage.setAttributes(baseAttributes, range: NSRange(location: 0, length: storage.length))

        // Apply markdown styles in correct order (most specific first)
        let codeBlockRanges = renderCodeBlocks(in: storage, text: text)
        renderHeaders(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderBold(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderItalic(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderInlineCode(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderLists(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderDividers(in: storage, text: text, excludingRanges: codeBlockRanges)
        renderLinks(in: storage, text: text, excludingRanges: codeBlockRanges)

        storage.endEditing()
    }

    // MARK: - Private Rendering Methods

    private func isRangeExcluded(_ range: NSRange, from excludedRanges: [NSRange]) -> Bool {
        excludedRanges.contains { excludedRange in
            NSIntersectionRange(range, excludedRange).length > 0
        }
    }

    private func renderHeaders(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // H1: # Header
        let h1Pattern = "^#\\s+(.+)$"
        if let regex = try? NSRegularExpression(pattern: h1Pattern, options: .anchorsMatchLines) {
            regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
                guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }

                // Style the # marker
                let hashRange = NSRange(location: match.range.location, length: 1)
                storage.addAttributes([
                    .foregroundColor: markerColor,
                    .font: NSFont.monospacedSystemFont(ofSize: style.bodyFontSize, weight: .regular)
                ], range: hashRange)

                // Style the header content
                if match.numberOfRanges >= 2 {
                    let contentRange = match.range(at: 1)
                    storage.addAttributes([
                        .font: NSFont.systemFont(ofSize: style.headerFontSize, weight: .bold),
                        .foregroundColor: headerColor
                    ], range: contentRange)
                }
            }
        }

        // H2+: ## Header, ### Header, etc.
        let h2Pattern = "^(#{2,6})\\s+(.+)$"
        if let regex = try? NSRegularExpression(pattern: h2Pattern, options: .anchorsMatchLines) {
            regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
                guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }

                // Style the ## markers
                if match.numberOfRanges >= 2 {
                    let hashRange = match.range(at: 1)
                    storage.addAttributes([
                        .foregroundColor: markerColor,
                        .font: NSFont.monospacedSystemFont(ofSize: style.bodyFontSize, weight: .regular)
                    ], range: hashRange)
                }

                // Style the header content
                if match.numberOfRanges >= 3 {
                    let contentRange = match.range(at: 2)
                    storage.addAttributes([
                        .font: NSFont.systemFont(ofSize: style.subHeaderFontSize, weight: .semibold),
                        .foregroundColor: headerColor
                    ], range: contentRange)
                }
            }
        }
    }

    private func renderBold(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // Match **text** but not ***text***
        let boldPattern = "(?<!\\*)\\*\\*(?!\\*)(.+?)(?<!\\*)\\*\\*(?!\\*)"
        guard let regex = try? NSRegularExpression(pattern: boldPattern, options: []) else { return }

        regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
            guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }
            guard match.numberOfRanges >= 2 else { return }

            let contentRange = match.range(at: 1)

            // Style the content as bold
            let boldColor = style == .analysis ? yellowColor : NSColor(white: 1.0, alpha: 1.0)
            storage.addAttributes([
                .font: NSFont.systemFont(ofSize: style.bodyFontSize, weight: .bold),
                .foregroundColor: boldColor
            ], range: contentRange)

            // Dim the ** markers
            let openRange = NSRange(location: match.range.location, length: 2)
            let closeRange = NSRange(location: contentRange.upperBound, length: 2)
            storage.addAttribute(.foregroundColor, value: markerColor, range: openRange)
            storage.addAttribute(.foregroundColor, value: markerColor, range: closeRange)
        }
    }

    private func renderItalic(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // Match *text* but not **text** (single asterisk only)
        let italicPattern = "(?<!\\*)\\*(?!\\*)(.+?)(?<!\\*)\\*(?!\\*)"
        guard let regex = try? NSRegularExpression(pattern: italicPattern, options: []) else { return }

        regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
            guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }
            guard match.numberOfRanges >= 2 else { return }

            let contentRange = match.range(at: 1)

            // Apply italic style
            if let existingFont = storage.attribute(.font, at: contentRange.location, effectiveRange: nil) as? NSFont {
                let italicFont = NSFontManager.shared.convert(existingFont, toHaveTrait: .italicFontMask)
                storage.addAttribute(.font, value: italicFont, range: contentRange)
            }

            // Dim the * markers
            let openRange = NSRange(location: match.range.location, length: 1)
            let closeRange = NSRange(location: contentRange.upperBound, length: 1)
            storage.addAttribute(.foregroundColor, value: markerColor, range: openRange)
            storage.addAttribute(.foregroundColor, value: markerColor, range: closeRange)
        }
    }

    private func renderCodeBlocks(in storage: NSTextStorage, text: String) -> [NSRange] {
        let codeBlockPattern = "```(\\w*)\\n([\\s\\S]*?)```"
        guard let regex = try? NSRegularExpression(pattern: codeBlockPattern, options: []) else { return [] }

        var codeBlockRanges: [NSRange] = []
        let matches = regex.matches(in: text, range: NSRange(text.startIndex..., in: text))

        for match in matches {
            codeBlockRanges.append(match.range)

            // Extract language and code content
            let nsString = text as NSString
            let language = match.numberOfRanges >= 2 ? nsString.substring(with: match.range(at: 1)) : ""
            let codeContent = match.numberOfRanges >= 3 ? nsString.substring(with: match.range(at: 2)) : ""

            // Apply background to entire block first
            storage.addAttribute(.backgroundColor, value: codeBackgroundColor, range: match.range)

            // Dim the opening ``` + language line
            if match.numberOfRanges >= 2 {
                let langRange = match.range(at: 1)
                let openingEnd = langRange.upperBound + 1
                let openingRange = NSRange(location: match.range.location, length: min(openingEnd - match.range.location, match.range.length))
                storage.addAttributes([
                    .foregroundColor: markerColor,
                    .font: NSFont.monospacedSystemFont(ofSize: style.codeFontSize, weight: .regular)
                ], range: openingRange)
            }

            // Apply syntax highlighting to code content
            if match.numberOfRanges >= 3 {
                let codeRange = match.range(at: 2)
                let highlighted = syntaxHighlighter.highlight(codeContent, language: language.isEmpty ? nil : language)

                // Apply highlighted attributes to the code range
                let offset = codeRange.location
                highlighted.enumerateAttributes(in: NSRange(location: 0, length: highlighted.length), options: []) { attrs, range, _ in
                    let targetRange = NSRange(location: offset + range.location, length: range.length)
                    if targetRange.upperBound <= storage.length {
                        storage.addAttributes(attrs, range: targetRange)
                        // Ensure background color is applied
                        storage.addAttribute(.backgroundColor, value: codeBackgroundColor, range: targetRange)
                    }
                }
            }

            // Dim the closing ```
            let closingStart = match.range.upperBound - 3
            if closingStart >= match.range.location {
                let closingRange = NSRange(location: closingStart, length: 3)
                storage.addAttributes([
                    .foregroundColor: markerColor,
                    .font: NSFont.monospacedSystemFont(ofSize: style.codeFontSize, weight: .regular)
                ], range: closingRange)
            }
        }

        return codeBlockRanges
    }

    private func renderInlineCode(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        let codePattern = "`([^`\\n]+)`"
        guard let regex = try? NSRegularExpression(pattern: codePattern, options: []) else { return }

        regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
            guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }

            // Style the entire match (including backticks)
            storage.addAttributes([
                .font: NSFont.monospacedSystemFont(ofSize: style.codeFontSize, weight: .medium),
                .foregroundColor: inlineCodeColor,
                .backgroundColor: inlineCodeBgColor
            ], range: match.range)

            // Dim the backticks
            let openRange = NSRange(location: match.range.location, length: 1)
            let closeRange = NSRange(location: match.range.upperBound - 1, length: 1)
            storage.addAttribute(.foregroundColor, value: markerColor, range: openRange)
            storage.addAttribute(.foregroundColor, value: markerColor, range: closeRange)
        }
    }

    private func renderLists(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // Match bullet lists (- item) and numbered lists (1. item)
        let listPattern = "^([ \\t]*)([-*+]|\\d+\\.)\\s"
        guard let regex = try? NSRegularExpression(pattern: listPattern, options: .anchorsMatchLines) else { return }

        regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
            guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }

            // Style the bullet/number marker in yellow
            if match.numberOfRanges >= 3 {
                let markerRange = match.range(at: 2)
                storage.addAttributes([
                    .foregroundColor: yellowColor,
                    .font: NSFont.systemFont(ofSize: style.bodyFontSize, weight: .semibold)
                ], range: markerRange)
            }
        }
    }

    private func renderDividers(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // Match horizontal rules: ---, ***, ___ (with optional spaces)
        let dividerPattern = "^[ \\t]*[-*_]{3,}[ \\t]*$"
        guard let regex = try? NSRegularExpression(pattern: dividerPattern, options: .anchorsMatchLines) else { return }

        // Process matches in reverse to avoid range shifting when replacing
        let matches = regex.matches(in: text, range: NSRange(text.startIndex..., in: text))
        for match in matches.reversed() {
            guard !isRangeExcluded(match.range, from: excludingRanges) else { continue }

            // Replace with a long horizontal line
            let dividerLine = "────────────────────────────────────────────────"
            let attributedDivider = NSAttributedString(string: dividerLine, attributes: [
                .foregroundColor: yellowColor.withAlphaComponent(0.4),
                .font: NSFont.systemFont(ofSize: style.bodyFontSize, weight: .light)
            ])
            storage.replaceCharacters(in: match.range, with: attributedDivider)
        }
    }

    private func renderLinks(in storage: NSTextStorage, text: String, excludingRanges: [NSRange]) {
        // Match [text](url) links
        let linkPattern = "\\[([^\\]]+)\\]\\(([^)]+)\\)"
        guard let regex = try? NSRegularExpression(pattern: linkPattern, options: []) else { return }

        regex.enumerateMatches(in: text, range: NSRange(text.startIndex..., in: text)) { match, _, _ in
            guard let match = match, !isRangeExcluded(match.range, from: excludingRanges) else { return }
            guard match.numberOfRanges >= 3 else { return }

            let textRange = match.range(at: 1)
            let urlRange = match.range(at: 2)

            // Style the link text
            storage.addAttributes([
                .foregroundColor: linkColor,
                .underlineStyle: NSUnderlineStyle.single.rawValue
            ], range: textRange)

            // Dim the brackets and URL
            let openBracket = NSRange(location: match.range.location, length: 1)
            let closeBracket = NSRange(location: textRange.upperBound, length: 1)
            let openParen = NSRange(location: urlRange.location - 1, length: 1)
            let closeParen = NSRange(location: urlRange.upperBound, length: 1)

            storage.addAttribute(.foregroundColor, value: markerColor, range: openBracket)
            storage.addAttribute(.foregroundColor, value: markerColor, range: closeBracket)
            storage.addAttribute(.foregroundColor, value: markerColor, range: openParen)
            storage.addAttribute(.foregroundColor, value: markerColor, range: closeParen)
            storage.addAttribute(.foregroundColor, value: markerColor.withAlphaComponent(0.4), range: urlRange)
        }
    }
}

File: Presentation/Styling/MarkdownRenderer.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text.RegularExpressions;

namespace InterviewMasterApp.Presentation.Styling
{
    /// <summary>
    /// Markdown renderer for plain text. Produces StyledText runs for UI or HTML output.
    /// Ported from Swift: MarkdownRenderer.swift
    /// </summary>
    public class MarkdownRenderer
    {
        public enum Style
        {
            Notes,
            Analysis
        }

        private readonly Style _style;
        private readonly SyntaxHighlighter _syntax;

        // Colors
        private readonly Color _yellow = Color.FromArgb(255, 255, 214, 0);
        private readonly Color _header = Color.FromArgb(255, 255, 214, 0);
        private readonly Color _code = Color.FromArgb(255, 229, 235, 242);
        private readonly Color _codeBg = Color.FromArgb(128, 31, 33, 38);
        private readonly Color _inlineCode = Color.FromArgb(255, 250, 184, 153);
        private readonly Color _inlineCodeBg = Color.FromArgb(204, 38, 38, 51);
        private readonly Color _link = Color.FromArgb(255, 102, 179, 255);
        private readonly Color _marker = Color.FromArgb(153, 128, 128, 128);

        private float HeaderSize => _style == Style.Notes ? 24 : 20;
        private float SubHeaderSize => _style == Style.Notes ? 20 : 18;
        private float BodySize => _style == Style.Notes ? 16 : 15;
        private float CodeSize => _style == Style.Notes ? 15 : 14;

        public MarkdownRenderer(Style style = Style.Notes)
        {
            _style = style;
            _syntax = new SyntaxHighlighter(CodeSize);
        }

        /// <summary>
        /// Render markdown into a StyledText structure.
        /// </summary>
        public StyledText Render(string text)
        {
            var styled = new StyledText();
            if (string.IsNullOrEmpty(text)) return styled;

            int idx = 0;
            while (idx < text.Length)
            {
                var start = text.IndexOf("```", idx, StringComparison.Ordinal);
                if (start < 0)
                {
                    RenderMarkdownSegment(styled, text.Substring(idx));
                    break;
                }

                // Render preceding segment
                if (start > idx)
                {
                    RenderMarkdownSegment(styled, text.Substring(idx, start - idx));
                }

                // Find end of code block
                var end = text.IndexOf("```", start + 3, StringComparison.Ordinal);
                if (end < 0)
                {
                    // Unclosed code block: render as plain
                    RenderMarkdownSegment(styled, text.Substring(start));
                    break;
                }

                // Extract language + code
                var headerEnd = text.IndexOf('\n', start + 3);
                string lang = string.Empty;
                string code = string.Empty;
                if (headerEnd > 0 && headerEnd < end)
                {
                    lang = text.Substring(start + 3, headerEnd - (start + 3)).Trim();
                    code = text.Substring(headerEnd + 1, end - (headerEnd + 1));
                }
                else
                {
                    code = text.Substring(start + 3, end - (start + 3));
                }

                // Opening line
                styled.Add("```" + lang + "\n", MarkerStyle());

                // Highlight code
                var highlighted = _syntax.Highlight(code, string.IsNullOrEmpty(lang) ? null : lang);
                foreach (var run in highlighted.Runs)
                {
                    var s = run.Style.Clone();
                    s.Background = _codeBg;
                    styled.Add(run.Text, s);
                }

                // Closing
                styled.Add("\n```", MarkerStyle());

                idx = end + 3;
            }

            return styled;
        }

        private void RenderMarkdownSegment(StyledText styled, string segment)
        {
            if (string.IsNullOrEmpty(segment)) return;

            var lines = segment.Split('\n');
            for (int i = 0; i < lines.Length; i++)
            {
                var line = lines[i];
                if (IsDivider(line))
                {
                    styled.Add(new string('-', 48) + "\n", new TextStyle { Foreground = Color.FromArgb(80, _yellow), FontSize = BodySize });
                    continue;
                }

                if (line.StartsWith("# "))
                {
                    styled.Add("# ", MarkerStyle());
                    styled.Add(line.Substring(2), new TextStyle { Foreground = _header, Bold = true, FontSize = HeaderSize });
                }
                else if (Regex.IsMatch(line, "^#{2,6}\\s"))
                {
                    var m = Regex.Match(line, "^(#{2,6})\\s(.*)$");
                    styled.Add(m.Groups[1].Value + " ", MarkerStyle());
                    styled.Add(m.Groups[2].Value, new TextStyle { Foreground = _header, Bold = true, FontSize = SubHeaderSize });
                }
                else
                {
                    RenderInline(styled, line);
                }

                if (i < lines.Length - 1) styled.Add("\n", BaseStyle());
            }
        }

        private void RenderInline(StyledText styled, string line)
        {
            if (string.IsNullOrEmpty(line)) return;

            var patterns = new List<(Regex regex, Func<Match, (string text, TextStyle style)> handler)>
            {
                // Links: [text](url)
                (new Regex("\\[([^\\]]+)\\]\\(([^)]+)\\)"), m =>
                {
                    var text = m.Groups[1].Value;
                    var url = m.Groups[2].Value;
                    var style = new TextStyle { Foreground = _link, Underline = true, FontSize = BodySize };
                    return (text, style);
                }),
                // Inline code: `code`
                (new Regex("`([^`\\n]+)`"), m =>
                {
                    var text = m.Groups[1].Value;
                    var style = new TextStyle { Foreground = _inlineCode, Background = _inlineCodeBg, Monospace = true, FontSize = CodeSize };
                    return (text, style);
                }),
                // Bold: **text**
                (new Regex("(?<!\\*)\\*\\*(?!\\*)(.+?)(?<!\\*)\\*\\*(?!\\*)"), m =>
                {
                    var text = m.Groups[1].Value;
                    var style = new TextStyle { Foreground = _style == Style.Analysis ? _yellow : Color.White, Bold = true, FontSize = BodySize };
                    return (text, style);
                }),
                // Italic: *text*
                (new Regex("(?<!\\*)\\*(?!\\*)(.+?)(?<!\\*)\\*(?!\\*)"), m =>
                {
                    var text = m.Groups[1].Value;
                    var style = new TextStyle { Italic = true, FontSize = BodySize };
                    return (text, style);
                })
            };

            int index = 0;
            while (index < line.Length)
            {
                Match? bestMatch = null;
                Regex? bestRegex = null;
                foreach (var (regex, _) in patterns)
                {
                    var m = regex.Match(line, index);
                    if (m.Success && (bestMatch == null || m.Index < bestMatch.Index))
                    {
                        bestMatch = m;
                        bestRegex = regex;
                    }
                }

                if (bestMatch == null)
                {
                    styled.Add(line.Substring(index), BaseStyle());
                    break;
                }

                // Add text before match
                if (bestMatch.Index > index)
                {
                    styled.Add(line.Substring(index, bestMatch.Index - index), BaseStyle());
                }

                // Add styled match
                var handler = patterns.First(p => p.regex == bestRegex).handler;
                var (text, style) = handler(bestMatch);
                styled.Add(text, style);

                index = bestMatch.Index + bestMatch.Length;
            }
        }

        private bool IsDivider(string line)
        {
            var trimmed = line.Trim();
            return Regex.IsMatch(trimmed, "^[-*_]{3,}$");
        }

        private TextStyle BaseStyle() => new TextStyle { Foreground = Color.FromArgb(242, 242, 242), FontSize = BodySize };
        private TextStyle MarkerStyle() => new TextStyle { Foreground = _marker, Monospace = true, FontSize = BodySize };
    }
}

